File: typescript/src/builders.ts
/**
 * Builder primitives for rulang (proposal #2).
 *
 * Mirrors `rulang/builders.py`. Consumers use these to construct ASTs
 * programmatically without regex-splitting rule strings. Pair with
 * `format()` for the text direction and `parse()` for the reverse.
 */

import type {
  Action,
  And,
  Binary,
  BinaryOp,
  Comparison,
  ComparisonOp,
  Compound,
  CompoundOp,
  Condition,
  Existence,
  Expr,
  FieldSegment,
  FunctionCall,
  IndexSegment,
  Literal,
  Not,
  NullCoalesce,
  NullSafeFieldSegment,
  Or,
  Path,
  PathRef,
  PathSegment,
  Return,
  Rule,
  SetAction,
  Unary,
  WorkflowCallAction,
  WorkflowCallExpr,
} from "./ast.js";

// --- paths --------------------------------------------------------------

export function pathOf(...parts: Array<string | number | Expr>): Path {
  if (parts.length === 0) throw new Error("pathOf() requires at least one part");
  if (parts.length === 1 && typeof parts[0] === "string" && (parts[0].includes(".") || parts[0].includes("?."))) {
    return parseDotted(parts[0]);
  }

  const root = parts[0];
  if (typeof root !== "string") {
    throw new TypeError(`path root must be a string, got ${typeof root}`);
  }

  const segments: PathSegment[] = [];
  for (let i = 1; i < parts.length; i++) {
    const part = parts[i];
    if (typeof part === "number") {
      segments.push({
        kind: "index",
        expr: { type: "literal", value: Math.trunc(part), literalType: "int" },
      } satisfies IndexSegment);
    } else if (typeof part === "string") {
      if (part.startsWith("?.")) {
        segments.push({ kind: "nullSafeField", name: part.slice(2) } satisfies NullSafeFieldSegment);
      } else {
        segments.push({ kind: "field", name: part } satisfies FieldSegment);
      }
    } else {
      segments.push({ kind: "index", expr: part } satisfies IndexSegment);
    }
  }
  return { type: "path", root, segments };
}

function parseDotted(text: string): Path {
  const segments: PathSegment[] = [];
  let i = 0;
  let rootEnd: number | null = null;
  while (i < text.length) {
    if (text[i] === ".") {
      rootEnd = i;
      break;
    }
    if (text[i] === "?" && text[i + 1] === ".") {
      rootEnd = i;
      break;
    }
    i += 1;
  }
  if (rootEnd === null) return { type: "path", root: text, segments: [] };
  const root = text.slice(0, rootEnd);
  i = rootEnd;
  while (i < text.length) {
    if (text.slice(i, i + 2) === "?.") {
      i += 2;
      let j = i;
      while (j < text.length && text[j] !== "." && text[j] !== "?") j += 1;
      segments.push({ kind: "nullSafeField", name: text.slice(i, j) });
      i = j;
    } else if (text[i] === ".") {
      i += 1;
      let j = i;
      while (j < text.length && text[j] !== "." && text[j] !== "?") j += 1;
      segments.push({ kind: "field", name: text.slice(i, j) });
      i = j;
    } else {
      break;
    }
  }
  return { type: "path", root, segments };
}

// --- literals / expressions --------------------------------------------

export function lit(value: unknown): Literal {
  if (value === null || value === undefined) return { type: "literal", value: null, literalType: "none" };
  if (typeof value === "boolean") return { type: "literal", value, literalType: "bool" };
  if (typeof value === "number") {
    return Number.isInteger(value)
      ? { type: "literal", value, literalType: "int" }
      : { type: "literal", value, literalType: "float" };
  }
  if (typeof value === "string") return { type: "literal", value, literalType: "string" };
  if (Array.isArray(value)) {
    return { type: "literal", value: value.map(asExpr), literalType: "list" };
  }
  throw new TypeError(`lit() does not accept values of type ${typeof value}`);
}

function asExpr(value: unknown): Expr {
  if (isAstExpr(value)) return value as Expr;
  if (isPath(value)) return { type: "pathRef", path: value as Path } satisfies PathRef;
  return lit(value);
}

function isAstExpr(v: unknown): boolean {
  if (!v || typeof v !== "object") return false;
  const t = (v as { type?: string }).type;
  return t === "literal" || t === "pathRef" || t === "functionCall" || t === "binary" ||
    t === "nullCoalesce" || t === "unary" || t === "workflowCallExpr";
}

function isPath(v: unknown): boolean {
  return !!v && typeof v === "object" && (v as { type?: string }).type === "path";
}

export function pathref(path: string | Path): PathRef {
  return { type: "pathRef", path: typeof path === "string" ? pathOf(path) : path };
}

export function fn(name: string, ...args: unknown[]): FunctionCall {
  return { type: "functionCall", name, args: args.map(asExpr) };
}

export function binary(left: unknown, op: BinaryOp, right: unknown): Binary {
  if (!["+", "-", "*", "/", "%"].includes(op)) {
    throw new Error(`invalid binary op: ${op}`);
  }
  return { type: "binary", left: asExpr(left), op, right: asExpr(right) };
}

export function unary(inner: unknown): Unary {
  return { type: "unary", op: "-", expr: asExpr(inner) };
}

export function nullCoalesce(left: unknown, right: unknown): NullCoalesce {
  return { type: "nullCoalesce", left: asExpr(left), right: asExpr(right) };
}

export function workflowExpr(name: string, ...args: unknown[]): WorkflowCallExpr {
  return { type: "workflowCallExpr", name, args: args.map(asExpr) };
}

// --- comparisons -------------------------------------------------------

function cmp(op: ComparisonOp, left: unknown, right: unknown): Comparison {
  return { type: "comparison", left: asExpr(left), op, right: asExpr(right) };
}

export const eq = (l: unknown, r: unknown) => cmp("==", l, r);
export const neq = (l: unknown, r: unknown) => cmp("!=", l, r);
export const lt = (l: unknown, r: unknown) => cmp("<", l, r);
export const gt = (l: unknown, r: unknown) => cmp(">", l, r);
export const lte = (l: unknown, r: unknown) => cmp("<=", l, r);
export const gte = (l: unknown, r: unknown) => cmp(">=", l, r);
export const inOp = (l: unknown, r: unknown) => cmp("in", l, r);
export const notIn = (l: unknown, r: unknown) => cmp("not in", l, r);
export const contains = (l: unknown, r: unknown) => cmp("contains", l, r);
export const notContains = (l: unknown, r: unknown) => cmp("not contains", l, r);
export const startswith = (l: unknown, r: unknown) => cmp("startswith", l, r);
export const endswith = (l: unknown, r: unknown) => cmp("endswith", l, r);
export const matches = (l: unknown, r: unknown) => cmp("matches", l, r);
export const containsAny = (l: unknown, r: unknown) => cmp("contains_any", l, r);
export const containsAll = (l: unknown, r: unknown) => cmp("contains_all", l, r);

export function exists(expr: unknown): Existence {
  return { type: "existence", expr: asExpr(expr), kind: "exists" };
}

export function isEmpty(expr: unknown): Existence {
  return { type: "existence", expr: asExpr(expr), kind: "is_empty" };
}

// --- logical -----------------------------------------------------------

export function and(...conditions: Condition[]): Condition {
  if (conditions.length < 2) throw new Error("and() requires at least two conditions");
  let result = conditions[0];
  for (let i = 1; i < conditions.length; i++) {
    result = { type: "and", left: result, right: conditions[i] } satisfies And;
  }
  return result;
}

export function or(...conditions: Condition[]): Condition {
  if (conditions.length < 2) throw new Error("or() requires at least two conditions");
  let result = conditions[0];
  for (let i = 1; i < conditions.length; i++) {
    result = { type: "or", left: result, right: conditions[i] } satisfies Or;
  }
  return result;
}

export function not(condition: Condition): Not {
  return { type: "not", inner: condition };
}

// --- actions -----------------------------------------------------------

function asPath(p: string | Path): Path {
  return typeof p === "string" ? pathOf(p) : p;
}

export function set_(path: string | Path, value: unknown): SetAction {
  return { type: "set", path: asPath(path), value: asExpr(value) };
}

export function compound(path: string | Path, op: CompoundOp, value: unknown): Compound {
  if (!["+=", "-=", "*=", "/="].includes(op)) throw new Error(`invalid compound op: ${op}`);
  return { type: "compound", path: asPath(path), op, value: asExpr(value) };
}

export function workflowCall(name: string, ...args: unknown[]): WorkflowCallAction {
  return { type: "workflowCallAction", name, args: args.map(asExpr) };
}

export function ret(value: unknown): Return {
  return { type: "return", value: asExpr(value) };
}

// --- rule --------------------------------------------------------------

export function rule(
  condition: Condition,
  actions: readonly Action[],
  options: { entityName?: string } = {},
): Rule {
  const entityName = options.entityName ?? "entity";
  const { reads, writes, workflowCalls } = extractReadsWrites(condition, actions, entityName);
  return {
    type: "rule",
    condition,
    actions,
    source: "",
    entityName,
    reads,
    writes,
    workflowCalls,
  };
}

// Re-derive reads/writes from the AST so builder-built rules carry the same
// extractor values as parsed rules.
function extractReadsWrites(
  condition: Condition,
  actions: readonly Action[],
  entityName: string,
): { reads: Set<string>; writes: Set<string>; workflowCalls: Set<string> } {
  const reads = new Set<string>();
  const writes = new Set<string>();
  const workflowCalls = new Set<string>();

  const pathStr = (p: Path): string => {
    const parts: string[] = [p.root];
    for (const s of p.segments) {
      if (s.kind === "field" || s.kind === "nullSafeField") parts.push(s.name);
      else parts.push("[*]");
    }
    if (parts[0] !== entityName) parts.unshift(entityName);
    return parts.join(".");
  };

  const collectReads = (e: Expr): void => {
    switch (e.type) {
      case "pathRef":
        reads.add(pathStr(e.path));
        for (const s of e.path.segments) if (s.kind === "index") collectReads(s.expr);
        return;
      case "functionCall":
        for (const a of e.args) collectReads(a);
        return;
      case "binary":
      case "nullCoalesce":
        collectReads(e.left);
        collectReads(e.right);
        return;
      case "unary":
        collectReads(e.expr);
        return;
      case "workflowCallExpr":
        workflowCalls.add(e.name);
        for (const a of e.args) collectReads(a);
        return;
      case "literal":
        if (e.literalType === "list") for (const item of e.value as Expr[]) collectReads(item);
        return;
    }
  };

  const walk = (c: Condition): void => {
    switch (c.type) {
      case "and":
      case "or":
        walk(c.left);
        walk(c.right);
        return;
      case "not":
        walk(c.inner);
        return;
      case "comparison":
        collectReads(c.left);
        collectReads(c.right);
        return;
      case "existence":
        collectReads(c.expr);
        return;
    }
  };

  walk(condition);

  for (const action of actions) {
    switch (action.type) {
      case "set":
        writes.add(pathStr(action.path));
        collectReads(action.value);
        for (const s of action.path.segments) if (s.kind === "index") collectReads(s.expr);
        break;
      case "compound": {
        const ps = pathStr(action.path);
        writes.add(ps);
        reads.add(ps);
        collectReads(action.value);
        for (const s of action.path.segments) if (s.kind === "index") collectReads(s.expr);
        break;
      }
      case "workflowCallAction":
        workflowCalls.add(action.name);
        for (const a of action.args) collectReads(a);
        break;
      case "return":
        collectReads(action.value);
        break;
    }
  }

  return { reads, writes, workflowCalls };
}

File: python/src/rulang/builders.py
"""
Builder primitives for rulang (proposal #2).

Ergonomic constructors so consumers (visual builders, rule generators, LLM
agents) can assemble ASTs programmatically without regex-splitting rule
strings. Combine with `format()` to go from structured input to canonical
text; use `parse()` for the reverse.

Round-trip guarantees (enforced by tests):

    parse(format(rule)) structurally equals rule
    format(parse(text)) is canonical

Example:

    from rulang.builders import rule, eq, path_of, lit, set_
    r = rule(
        condition=eq(path_of("entity.age"), lit(18)),
        actions=[set_(path_of("entity.adult"), lit(True))],
    )
    rulang.format(r)   # "entity.age == 18 => entity.adult = true"
"""

from __future__ import annotations

from typing import Any, Iterable, Union

from rulang.ast import (
    Action,
    And,
    Binary,
    BinaryOp,
    Comparison,
    ComparisonOp,
    Compound,
    CompoundOp,
    Condition,
    Existence,
    ExistenceKind,
    Expr,
    FieldSegment,
    FunctionCall,
    IndexSegment,
    Literal_,
    Not,
    NullCoalesce,
    NullSafeFieldSegment,
    Or,
    Path,
    PathRef,
    PathSegment,
    Return,
    Rule,
    Set_,
    Unary,
    WorkflowCallAction,
    WorkflowCallExpr,
    _extract_reads_writes,
)


# --- Paths --------------------------------------------------------------


def path_of(*parts: Union[str, int, Expr]) -> Path:
    """
    Build a `Path` from convenient inputs.

    Accepts a dotted string (`"entity.a.b"`) as a single argument, or a
    mixed sequence of parts:

        path_of("entity", "a", "b")
        path_of("entity.a.b")
        path_of("entity", "items", 0, "name")   # literal int index
        path_of("entity", "items", lit(0))      # expression index

    Null-safe segments are written with a leading `?` in the dotted form,
    e.g. `path_of("entity.a?.b")`.
    """
    if len(parts) == 1 and isinstance(parts[0], str) and ("." in parts[0] or "?." in parts[0]):
        return _parse_dotted(parts[0])

    if not parts:
        raise ValueError("path_of() requires at least one part")

    root = parts[0]
    if not isinstance(root, str):
        raise TypeError(f"path root must be a string, got {type(root).__name__}")

    segments: list[PathSegment] = []
    for part in parts[1:]:
        if isinstance(part, int):
            segments.append(IndexSegment(expr=Literal_(value=part, type="int", span=None)))
        elif isinstance(part, str):
            if part.startswith("?."):
                segments.append(NullSafeFieldSegment(name=part[2:]))
            else:
                segments.append(FieldSegment(name=part))
        else:
            # Assume Expr
            segments.append(IndexSegment(expr=part))
    return Path(root=root, segments=tuple(segments), span=None)


def _parse_dotted(text: str) -> Path:
    """Parse a dotted path string including null-safe segments."""
    # Split on `.` and `?.` while preserving which is which
    segments: list[PathSegment] = []
    i = 0
    root_end = None
    while i < len(text):
        if text[i] == ".":
            root_end = i
            break
        if i + 1 < len(text) and text[i : i + 2] == "?.":
            root_end = i
            break
        i += 1
    if root_end is None:
        return Path(root=text, segments=(), span=None)
    root = text[:root_end]
    i = root_end
    while i < len(text):
        if text[i : i + 2] == "?.":
            i += 2
            j = i
            while j < len(text) and text[j] not in (".", "?"):
                j += 1
            segments.append(NullSafeFieldSegment(name=text[i:j]))
            i = j
        elif text[i] == ".":
            i += 1
            j = i
            while j < len(text) and text[j] not in (".", "?"):
                j += 1
            segments.append(FieldSegment(name=text[i:j]))
            i = j
        else:
            # Shouldn't happen given the grammar of what we accept here
            break
    return Path(root=root, segments=tuple(segments), span=None)


# --- Literals / expressions --------------------------------------------


def float_lit(value: float) -> Literal_:
    """Explicit float literal. Use when you want a float representation regardless
    of whether the input value is a whole number.

    TypeScript can't distinguish `3.0` from `3` at runtime; Python can. Use
    `float_lit(3)` when you want float semantics and need cross-runtime parity.
    """
    return Literal_(value=float(value), type="float", span=None)


def int_lit(value: int) -> Literal_:
    """Explicit int literal."""
    return Literal_(value=int(value), type="int", span=None)


def lit(value: Any) -> Literal_:
    """
    Build a `Literal_` with the type inferred from the Python value.

    Supports None, bool, int, float, str, and list (of literals or Exprs).
    Use `float_lit()` / `int_lit()` for explicit control when the Python
    value is ambiguous or when you need strict cross-runtime type parity.
    """
    if value is None:
        return Literal_(value=None, type="none", span=None)
    if isinstance(value, bool):
        return Literal_(value=value, type="bool", span=None)
    if isinstance(value, int):
        return Literal_(value=value, type="int", span=None)
    if isinstance(value, float):
        return Literal_(value=value, type="float", span=None)
    if isinstance(value, str):
        return Literal_(value=value, type="string", span=None)
    if isinstance(value, (list, tuple)):
        # Inside a list literal, context is unambiguous (all elements are
        # literals), so strings can be auto-wrapped. AST nodes pass through.
        items = tuple(
            v if isinstance(v, (Literal_, PathRef, FunctionCall, Binary, NullCoalesce, Unary, WorkflowCallExpr)) else lit(v)
            for v in value
        )
        return Literal_(value=items, type="list", span=None)
    raise TypeError(f"lit() does not accept values of type {type(value).__name__}")


def _as_expr(value: Any) -> Expr:
    """Coerce raw Python values to an Expr via `lit()`; pass through AST nodes.

    Raw strings are rejected because they're ambiguous: the caller might mean
    either a string literal or a path. Use `lit('hello')` for literals and
    `pathref('entity.name')` for paths.
    """
    if isinstance(value, (Literal_, PathRef, FunctionCall, Binary, NullCoalesce, Unary, WorkflowCallExpr)):
        return value
    if isinstance(value, Path):
        return PathRef(path=value, span=None)
    if isinstance(value, str):
        raise TypeError(
            "Raw strings in expression position are ambiguous — they could "
            "mean a path or a string literal. Use lit(%r) for a string "
            "literal or pathref(%r) for a path." % (value, value)
        )
    return lit(value)


def pathref(path: Union[str, Path]) -> PathRef:
    """Wrap a path in a `PathRef` for use as an expression."""
    if isinstance(path, str):
        path = path_of(path)
    return PathRef(path=path, span=None)


def fn(name: str, *args: Any) -> FunctionCall:
    """Build a built-in function call: `fn("lower", pathref("entity.x"))`."""
    return FunctionCall(name=name, args=tuple(_as_expr(a) for a in args), span=None)


def binary(left: Any, op: BinaryOp, right: Any) -> Binary:
    """Build a `Binary` (arithmetic) node."""
    if op not in ("+", "-", "*", "/", "%"):
        raise ValueError(f"invalid binary op: {op!r}")
    return Binary(left=_as_expr(left), op=op, right=_as_expr(right), span=None)


def unary(inner: Any) -> Unary:
    """Build a `Unary(-)` node."""
    return Unary(op="-", expr=_as_expr(inner), span=None)


def null_coalesce(left: Any, right: Any) -> NullCoalesce:
    """Build a null-coalesce `left ?? right`."""
    return NullCoalesce(left=_as_expr(left), right=_as_expr(right), span=None)


def workflow_expr(name: str, *args: Any) -> WorkflowCallExpr:
    """Build a `workflow("name", ...)` in expression position."""
    return WorkflowCallExpr(name=name, args=tuple(_as_expr(a) for a in args), span=None)


# --- Conditions --------------------------------------------------------


def _cmp(op: ComparisonOp, left: Any, right: Any) -> Comparison:
    return Comparison(left=_as_expr(left), op=op, right=_as_expr(right), span=None)


def eq(left: Any, right: Any) -> Comparison:
    return _cmp("==", left, right)


def neq(left: Any, right: Any) -> Comparison:
    return _cmp("!=", left, right)


def lt(left: Any, right: Any) -> Comparison:
    return _cmp("<", left, right)


def gt(left: Any, right: Any) -> Comparison:
    return _cmp(">", left, right)


def lte(left: Any, right: Any) -> Comparison:
    return _cmp("<=", left, right)


def gte(left: Any, right: Any) -> Comparison:
    return _cmp(">=", left, right)


def in_(left: Any, right: Any) -> Comparison:
    return _cmp("in", left, right)


def not_in(left: Any, right: Any) -> Comparison:
    return _cmp("not in", left, right)


def contains(left: Any, right: Any) -> Comparison:
    return _cmp("contains", left, right)


def not_contains(left: Any, right: Any) -> Comparison:
    return _cmp("not contains", left, right)


def startswith(left: Any, right: Any) -> Comparison:
    return _cmp("startswith", left, right)


def endswith(left: Any, right: Any) -> Comparison:
    return _cmp("endswith", left, right)


def matches(left: Any, right: Any) -> Comparison:
    return _cmp("matches", left, right)


def contains_any(left: Any, right: Any) -> Comparison:
    return _cmp("contains_any", left, right)


def contains_all(left: Any, right: Any) -> Comparison:
    return _cmp("contains_all", left, right)


def exists(expr: Any) -> Existence:
    """`x exists` — true if x is non-null."""
    return Existence(expr=_as_expr(expr), kind="exists", span=None)


def is_empty(expr: Any) -> Existence:
    """`x is_empty` — true if x is null, empty string, or empty collection."""
    return Existence(expr=_as_expr(expr), kind="is_empty", span=None)


def and_(*conditions: Condition) -> Condition:
    """Left-associative `and` of two or more conditions."""
    if len(conditions) < 2:
        raise ValueError("and_() requires at least two conditions")
    result = conditions[0]
    for c in conditions[1:]:
        result = And(left=result, right=c, span=None)
    return result


def or_(*conditions: Condition) -> Condition:
    """Left-associative `or` of two or more conditions."""
    if len(conditions) < 2:
        raise ValueError("or_() requires at least two conditions")
    result = conditions[0]
    for c in conditions[1:]:
        result = Or(left=result, right=c, span=None)
    return result


def not_(condition: Condition) -> Not:
    return Not(inner=condition, span=None)


# --- Actions -----------------------------------------------------------


def set_(path: Union[str, Path], value: Any) -> Set_:
    """`path = value` — a Set action."""
    return Set_(path=_as_path(path), value=_as_expr(value), span=None)


def compound(path: Union[str, Path], op: CompoundOp, value: Any) -> Compound:
    """Compound assignment (`+=`, `-=`, `*=`, `/=`)."""
    if op not in ("+=", "-=", "*=", "/="):
        raise ValueError(f"invalid compound op: {op!r}")
    return Compound(path=_as_path(path), op=op, value=_as_expr(value), span=None)


def workflow_call(name: str, *args: Any) -> WorkflowCallAction:
    """`workflow("name", ...)` as an action."""
    return WorkflowCallAction(name=name, args=tuple(_as_expr(a) for a in args), span=None)


def ret(value: Any) -> Return:
    """`ret value` — return from the rule."""
    return Return(value=_as_expr(value), span=None)


def _as_path(p: Union[str, Path]) -> Path:
    if isinstance(p, str):
        return path_of(p)
    return p


# --- Rule --------------------------------------------------------------


def rule(
    condition: Condition,
    actions: Iterable[Action],
    *,
    entity_name: str = "entity",
) -> Rule:
    """
    Build a `Rule` from a condition and a sequence of actions.

    Reads/writes/workflow_calls are computed automatically from the AST so
    the resulting Rule is indistinguishable from one produced by `parse()`
    (modulo spans and `source`, which will be None / ""). Round-trip with
    `format()` produces a canonical string that re-parses to the same AST.
    """
    action_tuple = tuple(actions)
    reads, writes, workflow_calls = _extract_reads_writes(condition, action_tuple, entity_name)
    return Rule(
        condition=condition,
        actions=action_tuple,
        source="",
        entity_name=entity_name,
        span=None,
        reads=frozenset(reads),
        writes=frozenset(writes),
        workflow_calls=frozenset(workflow_calls),
    )
